app/src/main/java/com/example/gamingex/framework/impl/SingleTouchHandler.kt

```diff
@@ -0,0 +1,78 @@
+package com.example.gamingex.framework.impl
+
+import android.view.MotionEvent
+import android.view.View
+import com.example.gamingex.framework.Input.Companion.TouchEvent
+import com.example.gamingex.framework.TouchHandler
+
+
+class SingleTouchHandler(view: View, val scaleX: Float, val scaleY: Float): TouchHandler {
+    val factory = PoolTouchFactory()
+    val touchEventPool = Pool(factory, 100)
+    var isTouched = false
+    var touchX: Int = 0
+    var touchY: Int = 0
+    val touchEvents = ArrayList<TouchEvent>()
+    val touchEventsBuffer = ArrayList<TouchEvent>()
+
+    init {
+        view.setOnTouchListener(this)
+    }
+
+    override fun onTouch(v: View?, event: MotionEvent?): Boolean {
+        synchronized(this) {
```

Comment:
Since this function will usually be called from the main (UI) thread it is a bad idea to have it blocked when some other thread holds the monitor for i.e. `getTouchEvents`.

Reply:
The main and the UI thread are separate, you should check the FastRenderView, it starts the new Thread to display graphics

Reply:
@lundibundi Sorry, I misunderstood you. About the monitor, the games I planned to implement don`t rely so hard on the speed of input. The framework should be upgraded for more complex games. The main issues are input and graphics. For input I still don`t know how to upgraade it and for graphics I think Open GL ES would be good. To wipe all misunderstandings explain please, what did you mean under the UI thread: user interaction(Input, etc.) or the user interface (graphics and rendering)?

Reply:
And without `synchronised` the race condition would be possible. I am not good at multiThreading, so I don`t think I can rework it properly with my current knowledge

Reply:
The problem is not the speed of input, but the freezes of the UI which will happen if something blocks the UI thread.
The touches on Android are always handled on UI thread unless you delegate them yourself (which in turn would incur unnecessary latency and overhead).

> To wipe all misunderstandings explain please, what did you mean under the UI thread:

Oh, take a look at this article https://developer.android.com/guide/components/processes-and-threads#Threads, by the main thread I mean exactly what that article means
`the system creates a thread of execution for the application, called "main."... As such, the main thread is also sometimes called the UI thread`

> And without `synchronized` the race condition would be possible

This should go the other way, `synchronized` will not magically solve all multithreading issues. The issues need to first be identified and then thought about and solved one way or another.
In this case if you really need to store all events and deliver them in batch (which is not that common, usually you would only need to listen for new events which naturally solves this issue) then you have at least 2 options:
1. If the fact that currently processed event can be missed is acceptable - then use Java Thread Safe Collections (just google them) or just replace the array guarded with a mutex\monitor.
2. If you really want to deliver all events then you would need to somehow block the thread that wants to get `getTouchEvents` while the `onTouch` listener is processing new event. You could achieve this with a combination of `volatile` boolean flag and `wait()/notify()` sequence (from the top of my head, most likely there is a better solution).


>  so I don`t think I can rework it properly with my current knowledge

No worries, this is not a multithreading class so that's not gonna be an issue here.